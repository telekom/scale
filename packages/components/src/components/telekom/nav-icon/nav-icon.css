/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-nav-icon {
  --spacing-mobile: 0 6px;
  --font-size-mobile: var(--scl-font-size-10);
  --line-height-mobile: var(--scl-font-line-height-120);
  --font-weight-mobile: var(--scl-font-weight-bold);

  --spacing-desktop: 0 0 0 var(--scl-spacing-16);
  --font-size-desktop: var(--scl-font-size-12);
  --line-height-desktop: var(--scl-font-line-height-133);

  --color: var(--scl-color-primary);
  --color-hover: var(--scl-color-primary-hover);
  --color-active: var(--scl-color-primary-active);
}

.meta-navigation__item-link {
  color: var(--scl-color-text-standard);
  display: flex;
  transition: all 0.2s ease-in-out;
  align-items: center;
  font-weight: var(--font-weight-mobile);
  text-decoration: none;
}

.meta-navigation__item--selected .meta-navigation__item-link {
  color: var(--color);
}

.meta-navigation__item-link:hover {
  color: var(--color-hover);
}
.meta-navigation__item-link:active {
  color: var(--color-active);
}

@media (max-width: 1023px) {
  .meta-navigation__item-link {
    min-width: 24px;
  }
  .meta-navigation__item-link {
    margin: var(--spacing-mobile);
    font-size: var(--font-size-mobile);
    line-height: var(--line-height-mobile);
    flex-direction: column;
  }
  .meta-navigation__item.mobile-menu {
    width: 50px;
    text-align: center;
    cursor: pointer;
  }
  .meta-navigation__item-link .meta-navigation__item-link-icon {
    width: 18px;
    height: 18px;
    margin-bottom: 4px;
  }
}

@media (min-width: 1024px) {
  .meta-navigation__item scale-menu-flyout {
    height: 24px;
  }
  .meta-navigation__item-link {
    margin: var(--spacing-desktop);
    font-size: var(--font-size-desktop);
    line-height: var(--line-height-desktop);
  }
  .meta-navigation__item-link .meta-navigation__item-link-icon {
    margin-right: 6px;
  }
}
