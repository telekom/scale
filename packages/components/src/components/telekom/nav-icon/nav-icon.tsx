/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, Prop, Element, h } from '@stencil/core';
import classNames from 'classnames';
import { renderIcon } from '../../../utils/render-icon';
import statusNote from '../../../utils/status-note';

@Component({
  tag: 'scale-nav-icon',
  styleUrl: './nav-icon.css',
})
export class NavIcon {
  @Element() host: HTMLElement;
  /** (optional) if this item is active */
  // DEPRECATED - active should replace isActive
  @Prop() isActive: boolean;
  @Prop() active: boolean;
  /** (optional) href value */
  @Prop() href?: string = 'javascript:void(0);';
  @Prop() clickLink: any;
  @Prop() icon: string;
  // DEPRECATED - mobileMenuOpen should replace isMobileMenuOpen
  @Prop() isMobileMenuOpen?: boolean = false;
  @Prop() mobileMenuOpen?: boolean = false;
  @Prop() refMobileMenuToggle?: any;
  @Prop() refMobileUserMenuToggle?: any;
  @Prop() refUserMenuToggle?: any;

  componentWillRender() {
    // make sure the deprecated props overwrite the actual ones if used
    // and show status note deprecated
    if (this.isActive !== undefined) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "isActive" is deprecated. Please use the "active" property!',
        type: 'warn',
        source: this.host,
      });
    }
    if (this.isMobileMenuOpen !== false) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "isMobileMenuOpen" is deprecated. Please use the "mobileMenuOpen" property!',
        type: 'warn',
        source: this.host,
      });
    }
  }

  render() {
    return (
      <li class={this.getCssClassMap()}>
        <a
          class="meta-navigation__item-link"
          ref={
            this.refMobileMenuToggle ||
            this.refMobileUserMenuToggle ||
            this.refUserMenuToggle
          }
          href={this.href}
          onClick={this.clickLink}
          onKeyDown={(event) => {
            if (!this.refMobileMenuToggle) {
              return;
            }
            if (['Enter', ' ', 'Escape', 'Esc'].includes(event.key)) {
              event.preventDefault();
              this.clickLink(event);
            }
          }}
        >
          {renderIcon({
            tag: `scale-icon-${this.icon}`,
            attributes: {
              class: 'meta-navigation__item-link-icon',
              selected: this.active || this.isActive,
            },
          })}
          <span class="meta-navigation__item-label">
            <slot></slot>
          </span>
        </a>
      </li>
    );
  }

  getCssClassMap() {
    return classNames(
      'meta-navigation__item',
      (this.active ||
        this.isActive ||
        this.mobileMenuOpen ||
        this.isMobileMenuOpen) &&
        'meta-navigation__item--selected',
      !!this.refMobileMenuToggle && 'mobile-menu'
    );
  }
}
