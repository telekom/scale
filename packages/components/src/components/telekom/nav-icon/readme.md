# scale-nav-icon



<!-- Auto Generated Below -->


## Properties

| Property                  | Attribute                     | Description                       | Type      | Default                 |
| ------------------------- | ----------------------------- | --------------------------------- | --------- | ----------------------- |
| `active`                  | `active`                      |                                   | `boolean` | `undefined`             |
| `clickLink`               | `click-link`                  |                                   | `any`     | `undefined`             |
| `href`                    | `href`                        | (optional) href value             | `string`  | `'javascript:void(0);'` |
| `icon`                    | `icon`                        |                                   | `string`  | `undefined`             |
| `isActive`                | `is-active`                   | (optional) if this item is active | `boolean` | `undefined`             |
| `isMobileMenuOpen`        | `is-mobile-menu-open`         |                                   | `boolean` | `false`                 |
| `mobileMenuOpen`          | `mobile-menu-open`            |                                   | `boolean` | `false`                 |
| `refMobileMenuToggle`     | `ref-mobile-menu-toggle`      |                                   | `any`     | `undefined`             |
| `refMobileUserMenuToggle` | `ref-mobile-user-menu-toggle` |                                   | `any`     | `undefined`             |
| `refUserMenuToggle`       | `ref-user-menu-toggle`        |                                   | `any`     | `undefined`             |


## Dependencies

### Used by

 - [scale-app-header](../app-header)

### Depends on

- [scale-icon](../../icon)

### Graph
```mermaid
graph TD;
  scale-nav-icon --> scale-icon
  scale-app-header --> scale-nav-icon
  style scale-nav-icon fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
