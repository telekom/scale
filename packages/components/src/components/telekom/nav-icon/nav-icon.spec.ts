/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { NavIcon } from './nav-icon';
describe('nav-icon', () => {
  let page: SpecPage;
  beforeEach(async () => {
    page = await newSpecPage({
      components: [NavIcon],
      html: `<scale-nav-icon></scale-nav-icon>`,
    });
  });
  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });
  it('should reflect to set attributes/props', async () => {
    const specPage = await newSpecPage({
      components: [NavIcon],
      html: `<scale-nav-icon
                is-active="true"
                href="href"
                click-link="true"
                icon="icon"
                is-mobile-menu-open="true"
                ref-mobile-menu-toggle="">
              </scale-nav-icon>`,
    });
    expect(specPage.rootInstance.isActive).toBe(true);
    expect(specPage.rootInstance.href).toBe('href');
    expect(specPage.rootInstance.clickLink).toBe('true');
    expect(specPage.rootInstance.icon).toBe('icon');
    expect(specPage.rootInstance.isMobileMenuOpen).toBe(true);
    expect(specPage.rootInstance.refMobileMenuToggle).toBe('');
  });
  it('should trigger onClick()', async () => {
    page.root.clickLink = jest.fn();
    const anchor = page.root.querySelector(
      '.meta-navigation__item-link'
    ) as HTMLElement;
    await page.waitForChanges();
    anchor.click();
    await page.waitForChanges();
    expect(page.root.clickLink).toHaveBeenCalled();
  });
  it('should trigger onKeyDown', async () => {
    page.root.clickLink = jest.fn();
    page.root.refMobileMenuToggle = () => {};
    const anchor = page.root.querySelector(
      '.meta-navigation__item-link'
    ) as HTMLElement;
    await page.waitForChanges();
    anchor.dispatchEvent(
      new KeyboardEvent('keydown', {
        view: window,
        bubbles: true,
        cancelable: true,
        key: 'Escape',
      })
    );
    await page.waitForChanges();
    expect(page.root.clickLink).toHaveBeenCalled();
  });
  it('should trigger onClick()', async () => {
    page.root.clickLink = jest.fn();
    const anchor = page.root.querySelector(
      '.meta-navigation__item-link'
    ) as HTMLElement;
    await page.waitForChanges();
    anchor.dispatchEvent(
      new KeyboardEvent('keydown', {
        view: window,
        bubbles: true,
        cancelable: true,
        key: 'Escape',
      })
    );
    await page.waitForChanges();
    expect(page.root.clickLink).not.toHaveBeenCalled();
  });
});
