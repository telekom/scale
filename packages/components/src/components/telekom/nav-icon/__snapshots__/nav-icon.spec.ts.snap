// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nav-icon should match snapshot 1`] = `
<scale-nav-icon>
  <!---->
  <li class="meta-navigation__item">
    <a class="meta-navigation__item-link" href="javascript:void(0);">
      <scale-icon-undefined class="meta-navigation__item-link-icon"></scale-icon-undefined>
      <span class="meta-navigation__item-label"></span>
    </a>
  </li>
</scale-nav-icon>
`;
